repair/row_level.cc

```diff
@@ -676,6 +676,36 @@ void flush_rows(schema_ptr s, std::list<repair_row>& rows, lw_shared_ptr<repair_
     }
 }
 
+// For local reader: a permit is taken on the local shard.
+// For multi-shard reader: a permit is taken on each shard, smp::count permits
+// are taken in total.
+struct repair_reader_permit_meta {
+    std::vector<foreign_ptr<lw_shared_ptr<semaphore_units<>>>> permits{smp::count};
+};
+
+// If all_shards is set to true, permits on each shard are taken.
+// If all_shards is set to false, a single permit is taken from the specified shard.
+future<repair_reader_permit_meta> get_global_reader_permit(repair_service& rs, unsigned shard, bool all_shards) {
+    repair_reader_permit_meta meta;
+    // We need to serialize the process of taking permits. So the code to take
+    // the permits are performed on a single shard. The last shard is chosen as
+    // the coordinator shard.
+    co_await rs.container().invoke_on(smp::count -1, [&meta, shard, all_shards] (repair_service& rs) -> future<> {
+        co_await with_semaphore(rs.lock_sem(), 1, [&rs, &meta, shard, all_shards] () -> future<> {
+            co_await rs.container().invoke_on_all([&meta, shard, all_shards] (repair_service& rs) -> future<> {
```

Comment:
If all_shards is false, you don't need the invoke_on_all(), you can just `invoke_on(shard, ...)`